### ISSUE TITLE
Passing training=True to DNNRankingNetwork instance to invoke call() function

### ISSUE THREAD

**Issue by azagniotov** [tensorflow/ranking#316]
Hello Team,

Library versions:
- TensorFlow 2.5.0
- TensorFlow Ranking 0.4.2

## TL;DR:

I trained two models (let's calls them `A` and `B` (with fixed seed, same training dataset, the same hyper params). On each re-training run, each model consistently and deterministically produces the same results during training. The only difference is when the training the two models, in model `B`, I am passing `training=True` to `DNNRankingNetwork` instance (when invoking its call() function) when building an instance of `tf.keras.Model`. 

Because I pass in `training=True` when constructing model `B`, its training output consistently and deterministically different to the output of model `A`.

## Details

I noticed that in https://github.com/tensorflow/ranking/blob/v0.4.2/tensorflow_ranking/python/keras/model.py#L31-L78 when calling the instance of `DNNRankingNetwork` (it is actually an instance of `tf.keras.layers.Layer` through the inheritance chain) , you can pass another parameter `training=True`, e.g.: 
```
network(inputs=keras_inputs, mask=mask, training=True)
```
According to the docs in the code, this boolean arg controls whether running in training or inference mode. By default it is set to `None`.

I tried to trace what it is doing and how it is being used.  When we [pass the input of type TensorSpec](https://github.com/tensorflow/ranking/blob/v0.5.0/tensorflow_ranking/python/keras/model.py#L73) to `DNNRankingNetwork`, we actually invoke its `.call(..)` under the hood. This respectively is calling the [tensorflow_ranking/python/keras.EncodeListwiseFeatures.call(...)](https://github.com/tensorflow/ranking/blob/d325abcae6b22e3fd014e960a9b3aa35aa6d8732/tensorflow_ranking/python/keras/feature.py#L233-L260), which is calling the [tf.keras.layers.DenseFeatures.call(...)](https://github.com/keras-team/keras/blob/2f1cc1032ea51e5762ea4ed24e33deb33bc37075/keras/feature_column/dense_features.py#L139-L144)

By default, in the `DenseFeatures` , the `training` arg is None, which causes the invocation:
```
training = backend.learning_phase()
```
https://github.com/keras-team/keras/blob/2f1cc1032ea51e5762ea4ed24e33deb33bc37075/keras/backend.py#L311-L337

Eventually, in the `DenseFeatures`, the 'training' param gets passed to `FeatureColumn.get_dense_tensor(..)`

### Questions

It seems that passing `training=True` to `DNNRankingNetwork` instance (when invoking its call() function) affects the training output.

1. Why [model.py is not calling](https://github.com/tensorflow/ranking/blob/v0.5.0/tensorflow_ranking/python/keras/model.py#L31-L78) an instance `DNNRankingNetwork` with `training=True` when creating an instance of `tf.keras.Model`?
2. Should the  instance of `DNNRankingNetwork` be called with `training=True` in general, when creating a model instance for training

**Comment by franckjay** [comment]
Just curious as to the performance difference between Model A and B? Is there a hit to the holdout metrics?

**Comment by azagniotov** [comment]
Training model `A` without passing explicit `training=True` to the instance of `DNNRankingNetwork` layer when calling it. The following are epoch val loss output observed during training:
```
Epoch 00001: val_loss improved from inf to -0.83492
Epoch 00002: val_loss improved from -0.83492 to -0.84086
Epoch 00003: val_loss improved from -0.84086 to -0.84360
Epoch 00004: val_loss did not improve from -0.84360
Epoch 00005: val_loss improved from -0.84360 to -0.84488
Epoch 00006: val_loss improved from -0.84488 to -0.84637
...
...
Epoch 00010: val_loss did not improve from -0.84837
...
...
Epoch 00020: val_loss did not improve from -0.84837
...
...
Epoch 00050: val_loss did not improve from -0.84837
```

Now, I am training model `B`, by calling `DNNRankingNetwork` layer instance with an explicit `training=True`. The following are epoch output observed during training:
```
Epoch 00001: val_loss improved from inf to -0.83945
Epoch 00002: val_loss improved from -0.83945 to -0.83998
Epoch 00003: val_loss improved from -0.83998 to -0.84312
Epoch 00004: val_loss improved from -0.84312 to -0.84719
Epoch 00005: val_loss improved from -0.84719 to -0.84959
...
...
Epoch 00010: val_loss did not improve from -0.84959
...
...
Epoch 00020: val_loss did not improve from -0.85404
...
...
Epoch 00050: val_loss did not improve from -0.85860
```

The results are different JUST because I passed that argument in as `True` (To remind: both models have the same fixed seed, same training dataset input, the same hyper params)

I expected that explicitly setting `training=True` would NOT produce a different output, since we already should be in a training mode, hopefully correctly determined by the Framework

**Comment by azagniotov** [comment]
Hi @xuanhuiwang , if you could find time, could you please take a look? ^ 🙇🏼‍♂️

**Comment by ramakumar1729** [comment]
Hi @azagniotov , `model.fit()` method propagates `training=True` to all model sub-layers, so the difference in behavior is surprising. I'll take a closer look.

FYI we have deprecated RankingNetwork and related utilities for training and we are using native Keras training without feature columns but using Keras Preprocessing Layers instead. You can take a look at [https://www.tensorflow.org/ranking/tutorials/ranking_dnn_distributed](https://www.tensorflow.org/ranking/tutorials/ranking_dnn_distributed) for more information.
